### PR TITLE
[Spark] Fix stats with tightBounds check for AddFiles with deletionVectors

### PR DESCRIPTION
**Pull request by juliuszsompolski**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Check `DeletionVectorFilesHaveWideBoundsCheck` has been disabled for COMPUTE STATS because it reintroduces stats with tight bound to files with Deletion Vectors. However, there are other operations that can then copy over these AddFile actions with DVs and tight stats. These operations resulted in DELTA_ADDING_DELETION_VECTORS_WITH_TIGHT_BOUNDS_DISALLOWED error, which was a false positive.

In this PR we also attempt to introduce and discuss a "framework" for checks like that as a property of DeltaOperations, with DeltaOperations declaring as a member method whether a certain property and check should be performed. This is opposed to current practice, where many places in the code feature special cases like matching against a certain DeltaOperation and doing something special; this kind of code is very decentralized, and it's easy to miss if any new place or new operation needs such central handling. If this was centralized in DeltaOperations, this could lead to better discoverability of special cases and edge cases when implementing new operations.

## How was this patch tested?

Tests added.

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

**Comment by juliuszsompolski** [comment]
cc @bart-samwel @c27kwan

**Review by bart-samwel — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala [hunk 1]
```diff
@@ -72,6 +72,25 @@ object DeltaOperations {
...
+     * can then also recommit AddFile with deletionVector and tight bound stats that were recomputed
+     * before.
+     *
+     * An operation for which this can happen, and there is no way that it could be committing
+     * new deletion vectors, should set this to false to bypass this check.
+     * All other operations should set this to true, so that this is validated during commit.
+     *
+     * This is abstract to force the implementers of all operations to think about this setting.
+     * All operations should add a comment justifying this setting.
+     * Any operation that sets this to false should add a test in TightBoundsSuite.
```
[the remark this answers is not in the document]
Comment (juliuszsompolski) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:90]: I am afraid that has to be up to good reviewing practice...
I have found examples of some tests that e.g. use reflection to check the number of fields in some class and then comment there what needs to be done if a new one is added.
I can imagine using reflection to find all subclasses of DeltaOperation, listing the ones that already define this as false and have a test, and instruct in that test that if you add a new subclass that overrides it to false then you should add a test, but I think it's a bit overkill...
